src/auth/validate.rs

```diff
@@ -178,13 +180,33 @@ where
     Ok(possible_user)
 }
 
-pub fn extract_authorization_header(req: &HttpRequest) -> Result<&str, AuthenticationError> {
+pub fn extract_authorization_header(req: &HttpRequest) -> Result<String, AuthenticationError> {
     let headers = req.headers();
     let token_val: Option<&HeaderValue> = headers.get(AUTHORIZATION);
-    token_val
+    let val = token_val
         .ok_or_else(|| AuthenticationError::InvalidAuthMethod)?
         .to_str()
-        .map_err(|_| AuthenticationError::InvalidCredentials)
+        .map_err(|_| AuthenticationError::InvalidCredentials)?;
+
+    return match val.split_once(' ') {
+        Some(("Bearer", token)) => Ok(token.trim().to_string()),
+        Some(("Basic", token)) => {
+            let decoded = base64::engine::general_purpose::STANDARD
+                .decode(token.trim())
+                .map_err(|_| AuthenticationError::InvalidCredentials)?;
+
+            let credentials: String =
+                String::from_utf8(decoded).map_err(|_| AuthenticationError::InvalidCredentials)?;
+
+            Ok(credentials
+                .split_once(':')
+                .ok_or_else(|| AuthenticationError::InvalidCredentials)?
+                .1
+                .trim()
+                .to_string())
+        }
```

Comment:
It's kind of awkward and unexpected to have the client_id be in the request payload, and the secret extracted from the Basic authentication header. Having any of these be in the request body is not recommended as per the standard.

This would work for my generic implementation (since they're sent in both ways for compatibility reasons), but a client might expect to only send them through Basic authorization, or only through the HTTP payload.

I know it adds some complexity to the authentication routine, that's the blessing of standards that are 14 years old! 😇  My intuition would be to change the signature to output an enum, so a variant can hold 2 strings and pass them along to oauth checks.